### ISSUE TITLE
Track number sorting is incorrect

### ISSUE THREAD

**Issue by rendeko** [vanilla-music/vanilla#538]
Sort by is set to "Artist, album, track number" however the tracks are out of track number order. This is when looking at an album's songs. For example: Looking at the album "Hunky Dory" by David Bowie.

On my phone:
1. Quicksand (Demo Version) -- which is really tagged as track 14
2. Oh! You Pretty Things -- tagged as track 02
3. Changes -- tagged as track 01
4. Life On Mars? -- tagged as track 04
etc. 

I have copied the files back from my phone and verified that the track numbers are correct. The album and artist names are the same. The filenames are along the lines of "1.01 - Changes.mp3"  All of these files are v0 MP3's. I see no pattern between the way they are sorted on the phone.

This incorrect sorting occurs for all albums on my phone.

**Comment by adrian-bl** [comment]
Are the track numbers correctly picked up by the tag reader?

You can verify this by *long-pressing* on the info (eg: where Title / Artist is shown) in the playback view.

This will show additional information, including a 'Track' row

**Comment by rendeko** [comment]
Yes, Quicksand (Demo Version) says "track 14" and all the other songs show their correct track numbers.

**Comment by adrian-bl** [comment]
That's odd...

Sorting by artist, album, track actually does this internally:

> MediaLibrary.ContributorColumns.ARTIST_SORT+" 
> %1$s,"+MediaLibrary.AlbumColumns.ALBUM_SORT+" 
> %1$s,"+MediaLibrary.SongColumns.DISC_NUMBER+","
> +MediaLibrary.SongColumns.SONG_NUMBER,

So we are sorting by artist,album,disc_number,song_number .

I can only think of 2 possible issues that would lead to this situation:

* We wrote some random 'disc_number' into the DB, which is unlikely as we do not index this for MP3s
* The MP3 was tagged *after* it was indexed and we failed to re-index it - but in this case hitting 'Settings -> Media library -> Scan' should fix it (if the mtime changed)


Do you have root on your phone?
If yes, then please grab a copy of '/data/data/ch.blinkenlights.android.vanilla/databases/media-library.db' via

 adb pull /data/data/ch.blinkenlights.android.vanilla/databases/media-library.db

and send it to adrian@blinkenlights.ch

**Comment by rendeko** [comment]
Rescanning my library didn't work, including flushing it. In addition, I don't have root because I use CopperheadOS.

**Comment by adrian-bl** [comment]
Ok, could you then send me 2-3 sample MP3 files via email?

**Comment by adrian-bl** [comment]
Thanks for the sample files, i just tested it on my Nexus5 with LOS 14.1 and the files are indexed (and sorted) correctly:

```
sqlite> select song_num,disc_num,title from songs where path like '%.mp3' order by song_num;
song_num|disc_num|title
1|1|Changes
4|1|Life On Mars?
14|1|Quicksand (Demo Version)
```


So given the fact that your device also displays the track number, the issue itself doesn't seem to be related to the tag reader but something else strange must be going on :-/

> as all albums are sorted incorrectly, some using non-capitalized names where there should
> be capitalization

What do you mean with 'all albums are sorted incorrectly' ? Does even a simple thing like sorting by name fail? About the capitalization: Do these songs maybe have both ID3v2 and ID3v1?

> Would it be possible for me to change the location that the
> media-library.db is saved to somewhere accessable without root, and then
> recompile the apk to get it?

No, but you could change Vanilla Music itself to copy the database to eg. /sdcard
Another way to get a copy of the database would be if you had a root-enabled recovery which is able to mount your /data partition.

**Comment by rendeko** [comment]
What I meant by "all albums are sorted incorrectly" as in their songs are all sorted seemingly randomly, disregarding track number. However, while I was flushing the database to update the cover art I added to all my albums... the issue went away. Now every album's tracks go in order of track numbers. Unfortunately, this means I can't send you the broken .db file. Although there may not have been an issue in the first place that couldn't have been solved by flushing the database.

So this issue is closed, I guess. If anybody has this issue in the future, hopefully we can reopen and solve it.
